address_book.py

```diff
@@ -10,57 +10,68 @@ def __init__(self, name, email_address, phone_number):
 
     def show(self):
         """Shows details."""
-        print(f'Name = {self.name}, Email Address = {self.email_address}, Phone number = {self.phone_number}'
+        print(
+            f'Name = {self.name}, Email Address = {self.email_address}, Phone number = {self.phone_number}')
 
 
 class AddressBook:
-    def __init__(self):
-        self.contacts = {}
+    '''Write a name for your file as the argument for the AddressBook you are instantiating. 
+       This will be the file where the contacts are saved. .json will be added to the filename automatically.'''
+    # dont need to duplicate name
+
+    def __init__(self, contactsfile):
+        # This is where the data will be saved
+        self.contactsfile = contactsfile
+        self._contacts = {}
+        # if the file already exists, load the data in into the _contacts dict by iterating through the file
+        if Path(self.contactsfile).exists == True:
+            with open(self.contactsfile) as f:
+                data = json.load(f)
+
+            for name, details in data.items():
+                self._contacts[name] = Contact(
+                    name, details['email_address'], details['phone_number'])
+
+        elif Path(self.contactsfile).exists == False:
+            open(self.contactsfile)
 
     def add_contact(self, contact, object):
         """add contact to address book, contact is the key you will use to find the contact and
          the value is the contact object"""
-        self.contacts[contact] = object
+        self._contacts[contact] = vars(object)
 
         # Name of the file that will store the address book object
         # Write to the file
-        f = open(contactsfile, 'wb')
-        # Dump to the file
-        pickle.dump(self.contacts, f)
-        f.close()
+        with open(self.contactsfile, 'w') as f:
```

Comment:
You probably dont want to save to disk here. - You could do if you are really worried about say the computer crashing and data being lost, but I don't think you should. IO operations (read and write to hard disk) are really inefficient and normally you are happy to keep this stuff in RAM (ie just add it to self._contacts now). I think you should write another function called save() and then write the contacts file only when save is called. So a user will add a contact and or more and when the session is finished call save and save all changes in one go.